Replace a custom logger with off the shelf implementation

hubertp opened the pull request:
### Pull Request Description

This change replaces Enso's custom logger with an existing, mostly off the shelf logging implementation. The change attempts to provide a 1:1 replacement for the existing solution while requiring only a minimal logic for the initialization.

Loggers are configured completely via `logging-server` section in `application.conf` HOCON file, all initial logback configuration has been removed. This opens up a lot of interesting opportunities because we can benefit from all the well maintained slf4j implementations without being to them in terms of functionality.

Most important differences have been outlined in `docs/infrastructure/logging.md`.

### Important Notes

Addresses:
- #7253  
- #6739 

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [ ] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

hubertp committed:
Replace a custom logger with sth off the shelf

This change disables an Enso's custom logger with an existing, off the
shelf logging implementation with some custom setup and appenders.
The change attempts to provide a 1:1 replacement for the existing
solution.

Currently configuration is a mixture of `application.conf` HOCON file
with logback's configuration and programatically changed logger's
contexts.

hubertp committed:
format sbt

hubertp committed:
Fix native image build, many tests

wip

hubertp committed:
update changelog

hubertp committed:
minor formatting

hubertp committed:
fix test

hubertp committed:
Bring back jul's formatting to runtime

Conversion of raw messages to slf4j's style is now done in the handler.

JaroslavTulach commented on engine/runner/src/main/scala/org/enso/runner/DependencyPreinstaller.scala:
```diff
@@ -28,8 +30,8 @@ object DependencyPreinstaller {
-  def preinstallDependencies(projectRoot: File, logLevel: LogLevel): Unit = {
-    val logger = Logger[DependencyPreinstaller.type]
+  def preinstallDependencies(projectRoot: File, logLevel: Level): Unit = {
+    val logger = LoggerFactory.getLogger(classOf[DependencyPreinstaller.type])
```
This seems like _a usage of logger_. Does that mean we are supposed to use `org.slf4j` throughout the code base to do some logging?

JaroslavTulach commented on engine/runtime-language-epb/src/main/java/org/enso/interpreter/epb/EpbContext.java:
```diff
@@ -63,7 +63,7 @@ private static void initializeLanguages(
     }
     var log = environment.getLogger(EpbContext.class);
-    log.log(Level.FINE, "Initializing languages {0}", langs);
+    log.log(Level.FINE, "Initializing languages {}", langs);
```
Are you sure `{}` works? It is certainly not standard when using `LogRecord` standard formatter. As far as I know it didn't work before - as such I had to use `{0}`.

hubertp replied:
It _worked_ in a sense that all `LogRecord`s were forwarded to logback unformatted and logback recognized those. But when using plain jul this didn't work indeed (which was visible in tests). I reverted now and doing the `s/{\d+}/{}` before forwarding to logback.

JaroslavTulach commented on engine/runtime/src/main/scala/org/enso/compiler/Compiler.scala:
```diff
@@ -199,7 +199,7 @@ class Compiler(
             context.log(
               Level.SEVERE,
-              "Could not find entry point for compilation in package [{0}.{1}]",
+              "Could not find entry point for compilation in package [{}.{1}]",
```
Either `{}` works or `{1}` works... I assume it is unlikely both formats would work at the same time.

JaroslavTulach commented on engine/runtime/src/main/scala/org/enso/compiler/Compiler.scala:
```diff
@@ -199,7 +199,7 @@ class Compiler(
           case None =>
             context.log(
               Level.SEVERE,
```
This usage of logging is using `java.util.logging.Level` - are we supposed to use `java.util.logging.Logger` over `sl4j` in `engine/runtime`?

hubertp replied:
Given the size of this PR and the time spent I'm inclined to do this unification at a next stage. We still need to do a few things, like structural logging so this PR is not the end of the road. And we also don't degrade - this jul vs slf4j distinction in format has been here before.

JaroslavTulach commented on build.sbt:
```diff
@@ -880,6 +949,8 @@ lazy val `project-manager` = (project in file("lib/scala/project-manager"))
   .dependsOn(`polyglot-api`)
   .dependsOn(`runtime-version-manager`)
   .dependsOn(`library-manager`)
+  .dependsOn(`logging-utils-akka`)
```
Can we ensure this _implementation modules_ do not appear on compilation classpath of modules that are only supposed to use the logging? I'd like to avoid anyone using `ch.qos.logback` directly.

hubertp replied:
In `project-manager`? Tricky. I will see if `sbt` can support that request.

hubertp replied:
done

JaroslavTulach reviewed (commented):
It is not easy for me to see what is an API to use and what is an implementation. I'd like to ensure somehow the implementation classes that implement appenders/loghandlers do not appear on compilation path of modules that only use the logging.

I believe we need a guidance for using the logging. Is it `sl4j` or `jul`? What is the format of log record messages to use?

hubertp committed:
Fix some tests

Still not behaving as expected when setting up loggers for launcher.

hubertp committed:
Use test logger when collecting logs of a specific test

hubertp committed:
Bandaid to make sure launcher's logging works

When started, launcher should already make use of the logging's
configuration.

JaroslavTulach commented:
Important note. Currently the `slf4j` classes are **duplicated**. They are both in:
```
enso$ unzip -v runner.jar | grep -i slf4j
enso$ unzip -v runtime.jar | grep -i slf4j
```
that's wrong. We cannot have _two logging API classes_ and we cannot have _two logging implementations_!

hubertp committed:
Significantly simplifying logger's setup

hubertp commented:
> Important note. Currently the `slf4j` classes are **duplicated**. They are both in:
> 
> ```
> enso$ unzip -v runner.jar | grep -i slf4j
> enso$ unzip -v runtime.jar | grep -i slf4j
> ```
> 
> that's wrong. We cannot have _two logging API classes_ and we cannot have _two logging implementations_!

slf4j comes not only from direct libs dependencies that we control but also from indirect dependencies. The assembled jar will have slf4j API classes, no matter whether I turn everything in the repo to jul (I don't plan to) or not.
Runtime has no slf4j implementation, runner does. The former uses jul, the latter does not, that's all intended and works just fine.

hubertp committed:
More simplifications, documentation

hubertp committed:
Functionality-wise - done.

Cleaned up implementation for all appenders. All logback configs are
gone, everything is done via our application configs.

Still missing some documentation but no more features are planned.

hubertp committed:
logging-service - bye!

4e6 commented on engine/language-server/src/main/resources/application.conf:
```diff
@@ -11,13 +13,33 @@ akka {
+        hostname = ${?LOGSERVER_HOSTNAME}
+        hostname = "localhost"
+        port = ${?LOGSERVER_PORT}
+        port = 6000
```
I assume that the env var override should go after the default definition. Otherwise, you're overriding the env var with the value from config https://github.com/lightbend/config#optional-system-or-env-variable-overrides

hubertp replied:
Yes, thanks.

4e6 commented on engine/launcher/src/main/scala/org/enso/launcher/Launcher.scala:
```diff
@@ -207,7 +208,7 @@ case class Launcher(cliOptions: GlobalCLIOptions) {
     projectPath: Option[Path],
     versionOverride: Option[SemVer],
-    logLevel: LogLevel,
+    logLevel: Level,
```
I think having our own logging abstractions has benefits, i.e. whenever we want to change the implementation, the interface will stay the same. Do we really want to swap it with slf4j?

hubertp replied:
I didn't see any benefit of having a yet another layer of abstraction and slf4j appears to be a the most commonly used one. 
The implementation can be easily added (see e.g. `logging-truffle-connection`) for the existing example.

hubertp committed:
Ensure logging impl is not accidentally used

Ensured that imported subprojects don't add logback in transitive
dependencies and one has to explicitly enable it.

JaroslavTulach commented on lib/scala/logging-config/src/main/java/org/enso/logger/config/LoggerSetup.java:
```diff
@@ -0,0 +1,31 @@
+import org.slf4j.event.Level;
+
+/** Base classes to be implemented by the underlying loggging implementation. */
+public abstract class LoggerSetup {
```
`LoggerSetup` seems like a fine abstraction. I am assuming this `lib/scala/logging-config` doesn't depend on anything else than JDK and `slf4j`)...

All this class needs is `static LoggerSetup get()` method that will use `ServiceLoader` to lookup the `LogbackSetup` implementation once on runtime classpath.

hubertp replied:
Yeah, came to the same conclusion after our discussion. I didn't use `ServiceLoader` but some very basic custom reflection code instead. Seemed trivial enough to avoid another dependency. Should be fairly easy to switch now to `ServiceLoader` if you feel like there is any benefit of that.

JaroslavTulach replied:
> Yeah, came to the same conclusion after our discussion. 

OK, thank you.

> I didn't use `ServiceLoader` but some very basic custom reflection code instead... easy to switch now to 
> `ServiceLoader` if you feel like there is any benefit of that.

According to [Injectable singletons manifesto](http://wiki.apidesign.org/wiki/Injectable_Singleton) there are benefits over plain [reflection](https://github.com/enso-org/enso/pull/7559/commits/76c0efa3ca30ba844a3abe8d05ccfff4ae3a602b#diff-7f876974b7728ca4264314a6e7fac29592eecaa8f266296d5689ab8e25e4eae4R17) including _convention over configuration_, _testability_, reconfiguration by _assembling classpath_, etc. Moreover `ServiceLoader` support has been included into [module-info standard](https://docs.oracle.com/javase%2F9%2Fdocs%2Fapi%2F%2F/java/lang/module/ModuleDescriptor.Builder.html#provides-java.lang.String-java.util.List-). When/if/once we switch to _module path_ instead of _class path_, using `ServiceLoader` will be easier than your home made solution based on properties:
```java
Config c = ConfigFactory.load("enso-logging.conf");
String clazzName = c.getString(implClassKey);
Class<?> clazz = Class.forName(clazzName);
```
As far as I can tell your solution will probably reach the same goals as [Injectable singletons manifesto](http://wiki.apidesign.org/wiki/Injectable_Singleton) stated. The question to ask however is: Why invent our own solution when there is a standard one defined in the JDK?

_Btw: Wasn't the whole purpose of this PR to avoid home made solutions, when there are standard ones?_

hubertp replied:
OK. Done

JaroslavTulach commented on engine/launcher/src/main/scala/org/enso/launcher/cli/LauncherLogging.scala:
```diff
@@ -35,10 +30,9 @@ object LauncherLogging extends LoggingServiceSetupHelper {
-      Seq(stderrPrinter(colorMode, printExceptions = true))
-    )
+    val actualLogLevel = logLevel.getOrElse(defaultLogLevel)
+    LogbackSetup.get().setupConsoleAppender(actualLogLevel)
```
No (compile time) dependency on `LogbackSetup` please. Just use the [LoggerSetup.get static method](https://github.com/enso-org/enso/pull/7559/files#r1303229897) here.

hubertp committed:
Separate logger implementation from setup api

Logback implementation is now completely separate from the API used to
setup logging in different services. Loading is done once, by reading
implementation classes from the configuration file and intializing via
reflection.

Also discovered and fixed various problems with native image compilation.
More documentation that hopefully helps to clarify the usage.

hubertp committed:
Scala -> Java

hubertp committed:
Minor simplifications

hubertp committed:
update documentation

hubertp committed:
Tweaks to native image config

Caused by the previous moving of classes.

hubertp committed:
post-rebase tweaks

hubertp committed:
fmt

hubertp committed:
User ServiceLoader to load implementation

hubertp committed:
nits

JaroslavTulach commented on docs/infrastructure/logging.md:
```diff
@@ -10,250 +10,272 @@ order: 6
-components.
+one of the main components, allowing other components to connect to it. The
+service aggregates all logs in one place for easier analysis of the interaction
+between components. Components can also write to console, files directly without
```
```suggestion
between components. Components can also log to console or files directly without
```

JaroslavTulach replied:
Programs may of course print to console or write to files, but what does that have to do with logging? That was my first question after reading this sentence.

JaroslavTulach commented on docs/infrastructure/logging.md:
```diff
@@ -10,250 +10,272 @@ order: 6
+one of the main components, allowing other components to connect to it. The
+service aggregates all logs in one place for easier analysis of the interaction
+between components. Components can also write to console, files directly without
+involving an additional logging server.
```
```suggestion
involving the centralized logging service.
```

JaroslavTulach replied:
Refers to line 11 which mentions the "centrailized logging service".

JaroslavTulach commented on docs/infrastructure/logging.md:
```diff
@@ -10,250 +10,272 @@ order: 6
 
-The log level encoded as a number. Possible values are:
+During component's setup, its `application.conf` file is read, parsed and
+verified and available as part of the class hierarchy defined in
```
> verified and available

I don't understand.

JaroslavTulach commented on docs/infrastructure/logging.md:
```diff
@@ -10,250 +10,272 @@ order: 6
-Encodes an exception that is related to a log message.
+Config supports globs (`*`). For example, the config above sets
+`jdbc.JdbcBackend.statement` SQL statements logging to debug level, and the rest
+of the slick loggers to error level.
```
What is the FQN of the slick loggers? Is it `slick.jdbc.JdbcBackend.statement`? What is the difference between setting:
```
slick = error
slick.jdbc.JdbcBackend.statement = debug
```
and using the above `"*"` form?

hubertp replied:
Order matters. 
```
slick = error
slick.jdbc.JdbcBackend.statement = debug
```
will always  match `slick` classes.

I guess your question was more
```
  slick.jdbc.JdbcBackend.statement = debug
  slick = error
```
vs
```
  slick.jdbc.JdbcBackend.statement = debug
  slick.* = error
```

Well, no diff.

JaroslavTulach replied:
Then I don't understand why (it is important to stress that) _"Config supports globs (`*`)"_. If simple `slick = error` does the trick, then why advertise the complex syntax? Probably more important to stress that order matters.

JaroslavTulach commented on docs/infrastructure/logging.md:
```diff
@@ -10,250 +10,272 @@ order: 6
-that corresponds to the indicated stack location, for example `Main.java:123`.
-Native methods may be handled differently, as well as code from different
-languages, for example Enso also includes the columns - `Test.enso:4:3-19`.
+Any custom log level is therefore defined with `-Dx.y.Z.Logger.level` where `x`,
```
+1

JaroslavTulach commented on docs/infrastructure/logging.md:
```diff
@@ -10,250 +10,272 @@ order: 6
+Log output target is also configured in the `application.conf` files in the
+"appenders" section ("appender" is equivalent to `java.util.logging.Handler`
+semantics). Each appender section can provide further required and optional
+key/value pairs, to better customize the log target output.
```
At this point of document, I'd like to know how one chooses the each appender?

JaroslavTulach commented on docs/infrastructure/logging.md:
```diff
@@ -10,250 +10,272 @@ order: 6
+  with optional rolling file policy
+- [socket appender](#socket-appender) - appender that forwards log events to
+  some logging server
+- [sentry.io appender](#sentry-appender) - appender that forwards log events to
```
Good, support for `sentry.io` should be interesting for @PabloBuchu.

JaroslavTulach commented on docs/infrastructure/logging.md:
```diff
@@ -10,250 +10,272 @@ order: 6
+
+The two fields can be overridden via environment variables:
+
+- `hostanme` has an equivalent `$ENSO_LOGSERVER_HOSTNAME` variable
```
```suggestion
- `hostname` has an equivalent `$ENSO_LOGSERVER_HOSTNAME` variable
```

JaroslavTulach commented on docs/infrastructure/logging.md:
````diff
@@ -10,250 +10,272 @@ order: 6
+```
+  {
+    name = "sentry"
+    dsn = <string, required>
````
Shouldn't there be an environment variable to provide `dsn`? I assume @PabloBuchu would like to configure the value without changing engine/LS source code.

JaroslavTulach commented on docs/infrastructure/logging.md:
```diff
@@ -10,250 +10,272 @@ order: 6
-provided for the JVM.
+Previously, Enso came with its own implementation of Logger. That has been
+replaced with a mostly off-the-shelf logging implementation, avoiding
+maintenance costs for features that were already present in them.
```
There three sentences belong to a PR comment or description, but not to _"up-to-date"_ documentation. Who cares about _"previously"_?

JaroslavTulach commented on docs/infrastructure/logging.md:
```diff
@@ -10,250 +10,272 @@ order: 6
+globally during its execution. Currently, it is not possible to dynamically
+change where log events are being stored. The main class used for setting up
+logging is `org.enso.logger.LoggerSetup` but it should not be instantiated
+directly by the users. Instead, it should be retrieved with the Thread-safe
```
`LoggerSetup` is abstract. It cannot be instantiated. Btw. in current form the `get` method isn't thread-safe - it just pretends to be...

JaroslavTulach commented on lib/scala/logging-config/src/main/java/org/enso/logger/LoggerSetup.java:
```diff
@@ -0,0 +1,113 @@
+
+  public static LoggerSetup get() {
+    if (_instance == null) {
+      synchronized (lock) {
```
Double checked locking? That is an [antipattern](https://www.baeldung.com/java-singleton-double-checked-locking) as

> keep in mind with this pattern is that the field needs to be volatile to prevent cache incoherence issues.

I believe a simpler solution is to initialize `LoggerSetup` in class initializer. Btw. that might also fix problems with native-image as all the appendes could be pre-initiliazed before the image is generated.

JaroslavTulach replied:
Double checked locking is an antipattern. Initialize in static initializer instead.

JaroslavTulach commented on lib/scala/logging-logback/src/main/java/org/enso/logger/LogbackSetup.java:
```diff
@@ -0,0 +1,276 @@
+import org.slf4j.event.Level;
+
+@org.openide.util.lookup.ServiceProvider(service = LoggerSetup.class)
+public class LogbackSetup extends LoggerSetup {
```
`final`.

JaroslavTulach commented on lib/scala/logging-service/src/main/java/org/enso/logging/LoggingSetupHelper.java:
```diff
@@ -0,0 +1,163 @@
+ * Base class for any Enso service that needs to setup its logging.
+ *
+ * <p>Note: if this looks ugly and not very Java-friendly, it's because it is. It's a 1:1
+ * translation from Scala.
```
Yes, I can confirm it does look pretty ugly.

JaroslavTulach commented on lib/scala/logging-service/src/main/java/org/enso/logging/LoggingSetupHelper.java:
```diff
@@ -0,0 +1,163 @@
+   * @param logLevel maximal level of log events to be forwarded
+   * @param logMasking true if masking of sensitive data should be applied to all log messages
+   */
+  public void setupServerAndForwardLogs(Option<Level> logLevel, boolean logMasking)
```
Maybe just `Level logLevel` given the single usage currently passes in `Some(level)`.

JaroslavTulach commented on lib/scala/project-manager/src/main/scala/org/enso/projectmanager/boot/ProjectManager.scala:
```diff
@@ -256,25 +260,21 @@ object ProjectManager extends ZIOAppDefault with LazyLogging {
     ZIO
       .attempt {
-        Logging.setup(Some(level), None, colorMode, logMasking, profilingLog)
+        Logging.initLogger()
```
What's the point of separating `initLogger` and `setupServerAndForwardLogs` here? Why not hide it behind single `setup` method?

JaroslavTulach commented on docs/infrastructure/logging.md:
```diff
@@ -10,250 +10,272 @@ order: 6
+directories or connecting to an existing logging server once it starts accepting
+connections. That is why services don't call `LoggerSetup` directly but instead
+provide a service-specific implementation of
+`org.enso.logging.LoggingSetupHelper`. `LoggingSetupHelper` and `LoggerSetup`
```
Isn't there an inconsistency in naming? `Logger`Setup, but `Logging`SetupHelper. I'd expect the same prefix.

JaroslavTulach commented on lib/scala/logging-utils/src/test/scala/org/enso/logger/TestLogger.scala:
```diff
@@ -0,0 +1,29 @@
+    * @tparam A the type of logs to collect
+    * @return a tuple with the result of executing the closure and the list of log events collected
+    */
+  def gather[T, A](of: Class[A], action: => T): (T, List[TestLogMessage]) = {
```
Interesting class. Try to use it in `SerdeCompilerTest` instead of home made `MockHandler`. Maybe then turns out it'd be better if this class was written in Java.

JaroslavTulach commented on lib/scala/logging-jutil/src/main/java/org/enso/logger/Converter.java:
```diff
@@ -0,0 +1,48 @@
+
+public class Converter {
+
+  public static Level defaultLogLevel = TRACE;
```
`public static`, but not `final` field? E.g. singleton that can be changed? That's an anti pattern.

JaroslavTulach replied:
What's the purpose of `public` `static` non-`final` field?

JaroslavTulach commented on lib/scala/logging-jutil/src/main/java/org/enso/logger/JulHandler.java:
```diff
@@ -0,0 +1,63 @@
+import org.slf4j.LoggerFactory;
+
+/** java.util.logging.Handler that propagates all events to the equivalent SLF4J implementation. */
+public class JulHandler extends Handler {
```
`final`.

JaroslavTulach commented on lib/scala/logging-jutil/src/main/java/org/enso/logger/JulHandler.java:
```diff
@@ -0,0 +1,63 @@
+
+  private Formatter formattter;
+
+  public JulHandler() {
```
I think the constructor can be hidden and the class turned into singleton.

JaroslavTulach commented on lib/scala/logging-config/src/main/java/org/enso/logger/LoggerSetup.java:
```diff
@@ -0,0 +1,113 @@
+import org.slf4j.event.Level;
+
+/** Base class to be implemented by the underlying logging implementation. */
+public abstract class LoggerSetup {
```
`LoggerSetup` is used by "client code" - how can it be in the same project as implementation classes like `SocketAppender` that shall not be accessed directly?

There should be a module (maybe `logging-jutil`) that contains just the API to use by "client code". `Appender` class doesn't need to be visible by "client code", right?

JaroslavTulach commented on lib/scala/logging-utils/src/main/java/org/enso/logger/LoggerUtils.java:
```diff
@@ -0,0 +1,12 @@
+package org.enso.logger;
+
+public class LoggerUtils {
```
Isn't the code a bit _over-modularized_? Why do we have a `logging-utils` project and also `logging-jutil` project? Why `LoggerUtils` aren't merged with `Converter`?

JaroslavTulach commented on lib/scala/project-manager/src/main/resources/META-INF/native-image/org/enso/projectmanager/reflect-config.json:
```diff
@@ -565,5 +565,51 @@
+    ]
+  },
+  {
+    "name": "org.enso.logger.LogbackSetup",
```
Alternative to specifying reflection here is to initialize `LoggerSetup.get` singleton in static initializer. Then it will be "wired" before native image is generated.

JaroslavTulach reviewed (approved):
I left few comments, but I'd like us to also move forward. Merge soon!

radeusgd commented on engine/launcher/src/main/scala/org/enso/launcher/cli/GlobalCLIOptions.scala:
```diff
@@ -15,30 +14,27 @@ import org.enso.loggingservice.{ColorMode, LogLevel}
   autoConfirm: Boolean,
   hideProgress: Boolean,
   useJSON: Boolean,
-  colorMode: ColorMode,
```
I assume `colorMode` was only used in logging-related settings and it is no longer used?

What is the new behaviour of terminal colors? Is the logger output never colored now?

hubertp replied:
Coloring is not supported

4e6 commented on build.sbt:
```diff
@@ -338,24 +342,26 @@ lazy val enso = (project in file("."))
+val akkaTestkitTyped = akkaPkg("actor-testkit-typed") % Test
+val akkaHttp         = akkaHTTPPkg("http")
+val akkaSpray        = akkaHTTPPkg("http-spray-json")
+val akkaTest         = logbackPkg.map(_ % Test)
```
It should be called `logbackTest` right? It has nothing to do with akka.

4e6 commented on build.sbt:
```diff
@@ -679,8 +685,8 @@ lazy val `logging-utils` = project
-      "org.scalatest" %% "scalatest" % scalatestVersion % Test
-    )
+      "org.scalatest"    %% "scalatest" % scalatestVersion % Test
+    ) ++ logbackPkg.map(_ % Test)
```
the expression can be replaced with `akkaTest`

4e6 commented on build.sbt:
```diff
@@ -690,27 +696,64 @@ lazy val `logging-service` = project
+    version := "0.1",
+    libraryDependencies ++= Seq(
+      "org.slf4j"        % "slf4j-api"               % slf4jVersion,
+      "io.sentry"        % "sentry-logback"          % "6.28.0"           % Provided,
```
Why is it marked as `provided`? Where does it come from?

hubertp replied:
I wanted to provide support for sentry.io but not package it as a transitive dependency. Probably unnecessary when I think about it now.

4e6 commented on build.sbt:
```diff
@@ -690,27 +696,64 @@ lazy val `logging-service` = project
+    libraryDependencies ++= Seq(
+      "org.slf4j"        % "slf4j-api"               % slf4jVersion,
+      "io.sentry"        % "sentry-logback"          % "6.28.0"           % Provided,
+      "org.netbeans.api" % "org-openide-util-lookup" % netbeansApiVersion % "provided"
```
It seems this dependency can be removed

hubertp replied:
This is needed for `LogbackSetup` and `LogbackLoggingServiceFactory` service registration

4e6 commented on build.sbt:
```diff
@@ -719,10 +762,10 @@ lazy val filewatcher = project
+      "io.methvin"        % "directory-watcher" % directoryWatcherVersion,
+      "commons-io"        % "commons-io"        % commonsIoVersion,
+      "org.scalatest"    %% "scalatest"         % scalatestVersion % Test
+    ) ++ logbackPkg.map(_ % Test)
```
The expression can be replaced with `akkaTest` defined above

4e6 commented on build.sbt:
```diff
@@ -880,8 +923,11 @@ lazy val `project-manager` = (project in file("lib/scala/project-manager"))
+  .dependsOn(`logging-service`)
   .dependsOn(pkg)
   .dependsOn(`json-rpc-server`)
+  .dependsOn(`logging-logback` % Runtime)
```
Does it mean that the `logging-service` is an interface and the `logging-logback` is the implementation of the interface? I'd call it `logging-service-logback` then to be more clear

4e6 commented on build.sbt:
```diff
@@ -953,11 +999,10 @@ lazy val searcher = project
+      "com.typesafe.slick" %% "slick"       % slickVersion,
+      "org.xerial"          % "sqlite-jdbc" % sqliteVersion,
+      "org.scalatest"      %% "scalatest"   % scalatestVersion % Test
+    ) ++ logbackPkg.map(_ % Test)
```
Again, can be replaced with `akkaTest`

4e6 commented on engine/language-server/src/main/scala/org/enso/languageserver/util/UnhandledLogging.scala:
```diff
@@ -2,16 +2,18 @@ package org.enso.languageserver.util
 import com.typesafe.scalalogging.LazyLogging
-import org.enso.loggingservice.LogLevel
+import org.enso.logger.akka.AkkaConverter
+//import org.enso.loggingservice.LogLevel
```
cleanup

4e6 commented on engine/runtime/src/test/java/org/enso/interpreter/test/DebuggingEnsoTest.java:
```diff
@@ -55,6 +55,10 @@ public void initContext() {
         )
+        .option(
+                RuntimeOptions.LOG_LEVEL,
+                "FINEST"
```
I wonder how did it pass the formatted check

4e6 commented on lib/scala/logging-config/src/main/java/org/enso/logger/config/LoggersLevels.java:
```diff
@@ -0,0 +1,97 @@
+    Map<String, Level> fallbacks = new LinkedHashMap<>();
+    config
+        .entrySet()
+        .forEach(
```
You can call [`Map.forEach`](https://javadoc.scijava.org/Java17/java.base/java/util/Map.html#forEach(java.util.function.BiConsumer)) directly

hubertp replied:
Nope. `config` is not a map.

4e6 commented on lib/scala/logging-jutil/src/main/java/org/enso/logger/Converter.java:
```diff
@@ -0,0 +1,48 @@
+      case TRACE:
+        return java.util.logging.Level.FINEST;
+      default:
+        return java.util.logging.Level.ALL;
```
The match is already exhaustive. It should be a dead code, right?

hubertp replied:
We are dealing with Java ;)

JaroslavTulach replied:
Write the code like:
```java
    return switch (l) {
      case DEBUG -> -1;
      case ERROR -> -3;
      case INFO -> 5;
      case WARN -> 10;
      case TRACE -> 1;
    };
```
and then you will not need to complain.

4e6 commented on lib/scala/project-manager/src/test/scala/org/enso/projectmanager/infrastructure/languageserver/LanguageServerSupervisorSpec.scala:
```diff
@@ -106,6 +107,8 @@ class LanguageServerSupervisorSpec
 
   trait TestCtx {
 
+    LoggerSetup.get().setup()
```
Ok, given that it is a test.

4e6 reviewed (approved):
I left a bunch of comments but overall the PR looks good to me

radeusgd commented on docs/infrastructure/logging.md:
```diff
@@ -10,250 +10,272 @@ order: 6
-- 3 - indicating `DEBUG` level,
-- 4 - indicating `TRACE` level.
+As per [configuration schema](https://github.com/lightbend/config) any key can
+have a default value that can be overridden by an environmental variable. For
```
```suggestion
have a default value that can be overridden by an environment variable. For
```

nitpick, but I have never heard `environmental variable` before; sounds like a term related to ecology rather than programming 😅

radeusgd commented on docs/infrastructure/logging.md:
```diff
@@ -10,250 +10,272 @@ order: 6
+  a sentry.io service
 
-##### Parameters
+The appenders are defined by the `logging-service.appenders`. Currently only a
+single appender can be selected at a time. The selection may also be done via an
+environmental variable `$ENSO_DEFAULT_APPENDER`.
```
Would it be possible to still have an stdout appender when selecting the file appender? While file appender is very useful, I think it's good to be able to still see the logs in the terminal.

We can do `tail -f logfile` but allowing the app to do this automatically would be ideal.

hubertp replied:
I was going to add support for multiple appenders in the follow up PR. Also, it would probably make sense to have different levels for different loggers.

radeusgd commented on docs/infrastructure/logging.md:
```diff
@@ -10,250 +10,272 @@ order: 6
 
-Another example could be an info message (without attached exceptions):
+Rolling policy is a fully optional property of File Appender that would trigger
+automatic log rotation. All properties are optional with some reasonable
+defaults if not specified.
```
Can we describe what are these 'some reasonable defaults'? Otherwise, where can I find these defaults?

radeusgd commented on docs/infrastructure/logging.md:
```diff
@@ -281,51 +303,8 @@ String interpolation in log statements `s"Created $obj"` should be avoided
-Another useful tool is `TestLogger.gatherLogs` - a function that wraps an action
-and will return a sequence of logs reported when performing that action. It can
-be used to verify logs of an action inside of a test.
+The Logging Service provides a helper function `TestLogger.gatherLogs` that will
+execute the closure and collect all logs reported in the specified class. That
+way it can verify that all logs are being reported within the provided code.
```
❤️

radeusgd commented on lib/scala/logging-config/src/main/java/org/enso/logger/config/Appender.java:
```diff
@@ -0,0 +1,77 @@
+   * @param loggerSetup logger's setup to be used to be invoked with this appender
+   * @return true if logger has been setup correctly using this configuration, false otherwise
+   */
+  public Boolean setup(Level logLevel, LoggerSetup loggerSetup) {
```
```suggestion
  public boolean setup(Level logLevel, LoggerSetup loggerSetup) {
```

Can the return value ever be `null`? If not, let's un-box it.

radeusgd commented on lib/scala/logging-config/src/main/java/org/enso/logger/config/Appender.java:
```diff
@@ -0,0 +1,77 @@
+ * Base class for all appenders supported by Enso's logging configuration. Appenders determine what
+ * to do with the recorded log events
+ */
+public abstract class Appender {
```
Given that the contract of `LoggerSetup` limits the kinds of appenders that can be created, I think it would make sense to make `Appender` sealed, as there are no other implementations expected.

```suggestion
public sealed abstract class Appender permits ConsoleAppender, FileAppender, ... {
```

hubertp replied:
Fancy, I only see now that we started using this

radeusgd commented on lib/scala/logging-utils-akka/src/main/java/org/enso/logger/akka/AkkaConverter.java:
```diff
@@ -0,0 +1,55 @@
+  public static akka.event.Logging.LogLevel toAkka(Level level) {
+    switch (level) {
+      case ERROR:
+        return new akka.event.Logging.LogLevel(1); // Error
```
Could we try
```suggestion
        return new akka.event.Logging.ErrorLevel.asInt();
```
?
(I tried it in Scala, not sure if it will work OK in Java)

If possible, I'd prefer to use references from library rather than numeric constants + comments.

hubertp replied:
Won't work like that because `LogLevel` is a value class

radeusgd reviewed (approved):
As far as I have looked (did not go too deep), it looks good to me.

Just a few small suggestions from me

hubertp committed:
Address PR comments

hubertp committed:
linter

hubertp committed:
Apply suggestions from code review

Co-authored-by: Jaroslav Tulach <jaroslav.tulach@enso.org>
Co-authored-by: Radosław Waśko <radoslaw.wasko@enso.org>

hubertp committed:
Addressing more PR comments

hubertp committed:
Merge branch 'develop' into wip/hubert/7253-replace-custom-logger

hubertp committed:
Some Sentry additions, tested with prod

hubertp committed:
Fix native image build

hubertp committed:
formatting
